pkg/runtime/field_export_reconciler.go

```diff
@@ -0,0 +1,777 @@
+type fieldExportReconciler struct {
+	reconciler
+	// rd is only used if binding to an ACK resource (not `FieldExport`)
+	rd acktypes.AWSResourceDescriptor
+}
+
+// BindControllerManagerForFieldExport sets up the AWSResourceReconciler with an
+// instance of an upstream controller-runtime.Manager, watching for changes in
+// `FieldExport` CRs
+func (r *fieldExportReconciler) BindControllerManagerForFieldExport(mgr ctrlrt.Manager) error {
+	r.kc = mgr.GetClient()
+	r.apiReader = mgr.GetAPIReader()
+
+	if !ackcompare.IsNil(r.rd) {
+		return errors.New("cannot bind to field export. reconciler marked for reconciling AWS resources")
+	}
+
+	return ctrlrt.NewControllerManagedBy(
+		mgr,
+	).For(
+		// Read only field export objects
+		&ackv1alpha1.FieldExport{},
+	).WithEventFilter(
+		predicate.GenerationChangedPredicate{},
+	).Complete(r)
+}
+
+// BindControllerManagerForAWSResource sets up the AWSResourceReconciler with an
+// instance of an upstream controller-runtime.Manager, watching for changes in
+// AWS Resource CRs
+func (r *fieldExportReconciler) BindControllerManagerForAWSResource(mgr ctrlrt.Manager) error {
+	r.kc = mgr.GetClient()
+	r.apiReader = mgr.GetAPIReader()
+
+	if ackcompare.IsNil(r.rd) {
+		return errors.New("cannot bind to AWS resource. reconciler marked for reconciling field exports")
+	}
+
+	return ctrlrt.NewControllerManagedBy(
+		mgr,
+	).For(
+		r.rd.EmptyRuntimeObject(),
+	).WithEventFilter(
+		// Update on both status and spec changes
+		predicate.ResourceVersionChangedPredicate{},
+	).Complete(r)
+}
+
+// Reconcile implements `controller-runtime.Reconciler` and handles reconciling
+// a CR CRUD request
+func (r *fieldExportReconciler) Reconcile(ctx context.Context, req ctrlrt.Request) (ctrlrt.Result, error) {
+	return r.handleReconcileError(r.reconcile(ctx, req))
+}
+
+func (r *fieldExportReconciler) reconcile(ctx context.Context, req ctrlrt.Request) error {
+	// Determine if we are reconciling an ACK resource
+	if !ackcompare.IsNil(r.rd) {
+		return r.reconcileSourceResource(ctx, req)
+	}
+
+	// We are reconciling a field export CR
+	return r.reconcileFieldExport(ctx, req)
+}
+
+// reconcileFieldExport handles updates to `FieldExport` resources
+func (r *fieldExportReconciler) reconcileFieldExport(ctx context.Context, req ctrlrt.Request) error {
+	feObject, err := r.getFieldExport(ctx, req)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	sourceGK := feObject.Spec.From.Resource.GroupKind
+	sourceName := types.NamespacedName{
+		Name: *feObject.Spec.From.Resource.Name,
+		// We only support pulling from resources in
+		// the same namespace
+		Namespace: req.Namespace,
+	}
+
+	if feObject.DeletionTimestamp != nil {
+		return r.cleanup(ctx, feObject)
+	}
+
+	// Check if the target API group matches with the controller
+	var controllerRMF acktypes.AWSResourceManagerFactory
+	for _, v := range r.sc.GetResourceManagerFactories() {
+		controllerRMF = v
+		break
+	}
```

Comment:
Why do we need to do this when `fieldExportReconciler` contains an `rd` field which is the AWSResourceDescriptor that is the same as what is returned by `controllerRMF.ResourceDescriptor()` in the line below?

Reply:
> `fieldExportReconciler` contains an `rd` field

This is only true if we have bound the manager to an ACK resource. If we are binding to `FieldExport` then we leave that property as nil. In that case, this logic is the same as the adoption reconciler, where we simply fetch the first resource manager and use it to find the group that the controller is in charge of.